superwall/src/main/java/com/superwall/sdk/config/ConfigManager.kt

```diff
@@ -382,25 +383,30 @@ open class ConfigManager(
         oldConfig: Config,
         newConfig: Config,
     ) {
-        val oldPaywallIds = oldConfig.paywalls.map { it.identifier }.toSet()
-        val newPaywallIds = newConfig.paywalls.map { it.identifier }.toSet()
+        val oldPaywalls = oldConfig.paywalls
+        val newPaywalls = newConfig.paywalls
 
         val presentedPaywallId = paywallManager.currentView?.paywall?.identifier
-        val missingPaywallIds =
-            if (presentedPaywallId != null) {
-                oldPaywallIds.minus(presentedPaywallId).subtract(newPaywallIds)
-            } else {
-                oldPaywallIds.subtract(newPaywallIds)
-            }
-
-        missingPaywallIds.forEach {
-            val paywall = oldConfig.paywalls.first { wall -> wall.identifier == it }
-            val key =
-                PaywallCacheLogic.key(
-                    identifier = paywall.identifier,
-                    locale = factory.makeDeviceInfo().locale,
-                )
-            paywallManager.removePaywallView(key)
+        val oldPaywallCacheIds: Map<PaywallIdentifier, CacheKey> =
+            oldPaywalls
+                .map { it.identifier to it.cacheKey }
+                .toMap()
+        val newPaywallCacheIds: Map<PaywallIdentifier, CacheKey> = newPaywalls.map { it.identifier to it.cacheKey }.toMap()
+
+        val removedIds: Set<PaywallIdentifier> =
+            (oldPaywallCacheIds.keys - newPaywallCacheIds.keys).toSet()
+
+        val changedIds =
+            removedIds +
+                newPaywalls
+                    .filter {
+                        val oldCacheKey = oldPaywallCacheIds[it.identifier]
+                        val keyChanged = oldCacheKey != newPaywallCacheIds[it.identifier]
+                        oldCacheKey != null && keyChanged
+                    }.map { it.identifier } - presentedPaywallId
+
+        changedIds.toSet().filterNotNull().forEach {
+            paywallManager.removePaywallView(it)
```

Comment:
Here we're passing an identifier but the removePaywallView needs a cacheKey. On iOS I recreate the cache key with the identifier, but maybe you could just pass the cache key of the paywall directly from here

Reply:
The `paywallManager` forwards it to `cache` which turns the identifier into the cache key there, but with it being called `forKey` and `save` using the resolved `cacheKey` while `remove` uses non-resolved one is confusing tbh. 

Renamed and set the input as `PaywallIdentifier` to make it clearer, also made it so that both `save` and `remove` get identifiers passed in which then get resolved. Makes more sense 👍

Reply:
Ohhh got it. Another thing that confused me before was where we have a map of identifiers to cacheKey, and calling .keys on the map returns the identifiers. Not sure we can do anything about that but took me a second to wrap my head around it